neat_config.txt
## Template for gen_reads parallel
## Any parameter that is not required but has a default value will use the
## default value even if the variable is not included in the config. For
## required items, they must be included in the config and the must be given a value.
## All other items can be present or not. If present and the value is set to a single
## period, the variable will be treated as though it had been omitted. Please do
## not modify this template, but instead make a copy in your working directory. Done this
## way, you can run without even needing to declare -c.

# Absolute path to input reference fasta file
# type = string | required: yes
@reference = REQUIRED

# How to partition the reference for analysis. The default is just to partition the input into chromosomes.
# However, if you have only one chromosome, or multiple small contigs, or other siutations,
# you may see enhanced speed by subdividing the fasta contigs with the subdivision setting.
# type = string | required: yes | possible values: chrom, subdivision,
@partition_mode = subdivision

# Read length of the reads in the fastq output
# type = int | required: no | default = 101
@read_len = .

# Number of threads to request for NEAT. The recommended amount is the number of chromosomes in
# your input fasta plus 1. Any more than that will not improve performance much.
# type = int | required: no | default = 1
@threads = .

# Average Coverage for the entire genome.
# type = float | required: no | default = 10.0
@coverage = .

# Absolute path to file with sequencing error model
# type = string | required: no | default: <NEAT_DIR>/models/errorModel_default.pickle.gz
@error_model = .

# Average sequencing error rate for the sequencing machine
# type = float | required = no | must be between 0.0 and 0.3
@avg_seq_error = .

# This scales the quality scores to match the desired average sequencing error rate
# specified by @avg_seq_error.
# type: boolean | required = no | default = false
@rescale_qualities = .

# Desired ploidy
# type = int | required = no | default = 2
@ploidy = .

# Absolute path to vcf file containing variants that will always be included.
# type: string | required = no
@include_vcf = .

# Absolute path to bed file containing reference regions that the simulation
# should target.
# type = string | required = no
@target_bed = .

# Scalar value for coverage in regions outside the targeted bed. Example 0.5
# would get you roughly half the coverage as the on target areas.
# type: float | required = no | default = 0.0
@off_target_coverage = .

# Whether to discard areas outside the targeted bed region. By default, this is set
# to false and NEAT will use a different model for off-target regions but still
# include them in the final output.
# type: boolean | required = no | default = false
@discard_offtarget = .

# Absolute path to bed file containing reference regions that the simulation
# should discard.
# type = string | required = no
@discard_bed = .

# Absolute path to the mutation model pickle file. Omitting will ues the
# default model, built on human data.
# type: string | required = no | default: <NEAT_DIR>/models/errorModel_default.pickle.gz
@mutation_model = .

# Average mutation rate per base pair. Overall average is 0.001.
# This will rescale the rates determined by the model.
# type: float | required = no | must be between 0.0 and 0.3
@mutation_rate = .

# Absolute path to a bed file with mutation rates by region.
# Rates must be in the fourth column and be of the form "mut_rate=x.xx"
# Rates must be between 0.00 and 0.03
# type: string | required = no
@mutation_bed = .

# Cutoff score for changing a base-call to N.  Below the cutoff, base will be changed
# to N. If omitted, all bases will be shown, regardless of quality score
# type: int | required = no
@n_cutoff = .

# Absolute path to GC content model generated by compute_gc.py
# type: string | required = no | default: <NEAT_DIR>/models/gcBias_default.pickle.gz
@gc_model = .

# Whether the output should be paired ended. For certain conditions (i.e., vcf only or
# fasta only), this will be ignored. If this is true, then there must be an included fragment
# length model output from compute_fraglen.py or a mean and standard deviation (both).
# type: boolean | required = no | default = false
@paired_ended = .

# Absolute path to a pickle file containing the fragment length model output
# from compute_fraglen.py
# type: string | required = no
@fragment_model = .

# Mean for the paired end fragment length. This only applies if paired-ended is set to true.
# This number will form the mean for the sample distribution of the fragment lengths in the simulation
# Note: This number is REQUIRED if paired_ended is set to true, unless a fragment length model is used.
# type: float | required: no (unless paired-ended)
@fragment_mean = .

# Standard deviation for the paired end fragment length. This only applies if paired-ended is set to true.
# This number will form the standard deviation about the mean specified above for the sample distribution
# of the fragment lengths in the simulation.
# Note: This number is REQUIRED if paired_ended is set to true, unless a fragment length model is used.
# type: float | required: no (unless paired-ended)
@fragment_st_dev = .

# Whether to produce the golden bam file. This file will contain the reads
# aligned with the exact region of the genome
# type: boolean | required = no | default = false
@produce_bam = .

# Whether to produce a vcf file containing all the mutation errors added
# by NEAT.
# type: boolean | required = no | default = false
@produce_vcf = .

# Whether to output the mutated fasta. This will be the same information as the
# fastq, but just arranged as a fasta. Not compatible with paired-ended mode.
# NOTE: in Version 3.0 this will turn off fastq production. In future versions,
# you will be able to produce both, if desired.
# type: boolean | required = no | default = false
@produce_fasta = .

# Whether to output the fastq(s) of the reads. This is the default output. NEAT
# will produce 1 fastq for single ended reads or 2 fastqs for paired ended.
# type: boolean | required = no | default = true
@produce_fastq = .

# If set to true, this will ignore statistical models and force coverage to be
# constant across the genome. This is considered a debugging feature.
# type: boolean | required: no | default. false
@force_coverage = .

# Absolute path for a dir in which to create temp files for analysis. Ideally a dir with plenty of storage
# capacity and fast IO to the worker nodes. This will be deleted at the end of analysis.
# type: string | required: no | default: current working dir
@temp_dir = .

# Turn on debug mode by setting this to true. Debug mode will print certain messages
# that may help you pinpoint the problem.
# type: boolean | required: no | default: false
@debug = .

# Set an RNG seed value. Runs using idetical RNG values should produce identical results
# so things like read locations, variant positions, error positions, etc. should be the same.
# Useful for debugging.
# type: int | required: no
@rng_value = .

template_neat_config.txt
# Template for gen_reads parallel

# Any parameter that is not marked as required but has a default value will use the
# default value even if the variable is not included in the config. For
# required items, they must be included in the config and the must be given a value.
# All other items can be present or not. If present and the value is set to a single
# period, the variable will be treated as though it had been omitted. Please do
# not modify this template, but instead make a copy in your working directory. Done this
# way, you can run without even needing to declare -c.

# To enable fast editing: all lines starting with a single # are comments lines starting with a
# double ## are variable info, and lines starting with and @ sign are variables. You can quickly
# make a list of just the variables with, e.g., cat neat_config.txt | grep -v "#"

Use = sign as a
# separator. No quotation marks required around strings.
# Types:
#     - string - all ascii characters allowed, should be of a type consistent with the variable,
#                i.e., @reference requires a file path. Python reads all input as strings by default,
#                so quotation marks are not required.
#     - int - Python will attempt to cast this as an int datatype, so it must be consistent with Python int
#     - float - Python will attempt to cast this as a float, so it must be consistent with Python datatype float
#     - boolean - set to either true or false, case-insensitive.

# Absolute path to input reference fasta file
## type = string | required: yes
@reference = REQUIRED

# How to partition the reference for analysis. By default, NEAT will
# attempt to process one contig per thread. However, if you have very
# large fasta files, you will see additional runtime benefit from choosing
# the subdivision method, which will split the contigs up into equal sizes
# for processing. If you need further speedups and have access to a distributed system
# you can use a shell script wrapper around NEAT to split the fasta into
# contigs, then join the results later. NEAT does not feature translocations, so
# this will not affect NEAT's output.
## type = string | required: yes | possible values: chrom, subdivision,
@partition_mode = subdivision

# Read length of the reads in the fastq output
## type = int | required: no | default = 101
@read_len = .

# By default, temporary files will be stored in the same directory as the run directory
# All temporary files are cleaned up at the end. If you would like to specify a different
# location to store temp files, you can use this option.
## type = string | required: no | default = same as run directory
@temp_dir = .

# Number of threads to request for NEAT. Set this as high as you can for faster processing.
## type = int | required: no | default = 1
@threads = .

# Average Coverage for the entire genome.
## type = float | required: no | default = 10.0
@coverage = .

# Absolute path to file with sequencing error model
## type = string | required: no | default: <NEAT_DIR>/models/errorModel_default.pickle.gz
@error_model = .

# Average sequencing error rate for the sequencing machine
## type = float | required = no | must be between 0.0 and 0.3
@avg_seq_error = .

# This scales the quality scores to match the desired average sequencing error rate
# specified by @avg_seq_error.
## type: boolean | required = no | default = false
@rescale_qualities = .

# Desired ploidy
## type = int | required = no | default = 2
@ploidy = .

# Absolute path to vcf file containing variants that will always be included.
## type: string | required = no
@include_vcf = .

# Absolute path to bed file containing reference regions that the simulation
# should target.
## type = string | required = no
@target_bed = .

# Scalar value for coverage in regions outside the targeted bed. Example 0.5
# would get you roughly half the coverage as the on target areas.
## type: float | required = no | default = 0.0
@off_target_coverage = .

# Whether to discard areas outside the targeted bed region. By default, this is set
# to false and NEAT will use a different model for off-target regions but still
# include them in the final output.
## type: boolean | required = no | default = false
@discard_offtarget = .

# Absolute path to bed file containing reference regions that the simulation
# should discard.
## type = string | required = no
@discard_bed = .

# Absolute path to the mutation model pickle file. Omitting will ues the
# default model, built on human data.
## type: string | required = no | default: <NEAT_DIR>/models/errorModel_default.pickle.gz
@mutation_model = .

# Average mutation rate per base pair. Overall average is 0.001.
# This will rescale the rates determined by the model.
## type: float | required = no | must be between 0.0 and 0.3
@mutation_rate = .

# Absolute path to a bed file with mutation rates by region.
# Rates must be in the fourth column and be of the form "mut_rate=x.xx"
# Rates must be between 0.00 and 0.03
## type: string | required = no
@mutation_bed = .

# Cutoff score for changing a base-call to N.  Below the cutoff, base will be changed
# to N. If omitted, all bases will be shown, regardless of quality score
## type: int | required = no
@n_cutoff = .

# Absolute path to GC content model generated by compute_gc.py
## type: string | required = no | default: <NEAT_DIR>/models/gcBias_default.pickle.gz
@gc_model = .

# Whether the output should be paired ended. For certain conditions (i.e., vcf only or
# fasta only), this will be ignored. If this is true, then there must be an included fragment
# length model output from compute_fraglen.py or a mean and standard deviation (both).
## type: boolean | required = no | default = false
@paired_ended = .

# Absolute path to a pickle file containing the fragment length model output
# from compute_fraglen.py
## type: string | required = no
@fragment_model = .

# Mean for the paired end fragment length. This only applies if paired-ended is set to true.
# This number will form the mean for the sample distribution of the fragment lengths in the simulation
# Note: This number is REQUIRED if paired_ended is set to true, unless a fragment length model is used.
## type: float | required: no (unless paired-ended)
@fragment_mean = .

# Standard deviation for the paired end fragment length. This only applies if paired-ended is set to true.
# This number will form the standard deviation about the mean specified above for the sample distribution
# of the fragment lengths in the simulation.
# Note: This number is REQUIRED if paired_ended is set to true, unless a fragment length model is used.
## type: float | required: no (unless paired-ended)
@fragment_st_dev = .

# Whether to produce the golden bam file. This file will contain the reads
# aligned with the exact region of the genome
## type: boolean | required = no | default = false
@produce_bam = .

# Whether to produce a vcf file containing all the mutation errors added
# by NEAT.
## type: boolean | required = no | default = false
@produce_vcf = .

# Whether to output the mutated fasta. This will be the same information as the
# fastq, but just arranged as a fasta. Not compatible with paired-ended mode.
# NOTE: in Version 3.0 this will turn off fastq production. In future versions,
# you will be able to produce both, if desired.
## type: boolean | required = no | default = false
@produce_fasta = .

# Whether to output the fastq(s) of the reads. This is the default output. NEAT
# will produce 1 fastq for single ended reads or 2 fastqs for paired ended.
## type: boolean | required = no | default = true
@produce_fastq = .

# If set to true, this will ignore statistical models and force coverage to be
# constant across the genome. This is considered a debugging feature.
## type: boolean | required: no | default. false
@force_coverage = .

# Absolute path for a dir in which to create temp files for analysis. Ideally a dir with plenty of storage
# capacity and fast IO to the worker nodes. This will be deleted at the end of analysis.
## type: string | required: no | default: current working dir
@temp_dir = .

# Turn on debug mode by setting this to true. Debug mode will print certain messages
# that may help you pinpoint the problem.
## type: boolean | required: no | default: false
@debug = .

# Set an RNG seed value. Runs using idetical RNG values should produce identical results
# so things like read locations, variant positions, error positions, etc. should be the same.
# Useful for debugging.
## type: int | required: no
@rng_value = .

ChangeLog.md
# NEAT v3.0
- NEAT gen_reads now runs in Python 3 exclusively. The previous, Python 2 version is stored in the repo as v2.0, but will not be undergoing active development.
- Converted sequence objects to Biopython Sequence objects to take advantage of the Biopython library
- Converted cigar strings to lists. Now there are simply a functions that convert a generic cigar string to a list and vice versa.
- Tried to take better advantage of some Biopython libraries, such as for parsing fastas.
- For now, we've eliminated the "Jobs" option and the merge jobs function. We plan to implement multi-threading instead as a way to speed up NEAT's simulation process.
- Added a basic bacterial wrapper that will simulate multiple generations of bacteria based on an input fasta, mutate them and then produce the fastqs, bams, and vcfs for the resultant bacterial population.

# NEAT v3.1

Bug Fixes:
- Fixed an issue where NEAT could not read in gzipped vcf files
- Fixed a bug where NEAT was assuming a ploidy of 2
- Fixed a bug that was causing NEAT to produce exactly the same reverse strands as forward strands in paired ended mode
- Fixed a bug when reading in some mutation models.
- Added gzipping to all mutation models to maximize their compression
- Fixed some issues with vcf_compare

For improvements, we have a lot of great ideas for general improvements aimed at better simulating bacteria, but we believe this same improvements will have applications in other species as well. 
- Multiploidy - all right this has nothing to do with bacteria specifically, but it is a feature we would like to implement into gen_reads.
- Structural Variants - model large scale structural variants with an eye toward intergenic SVs.
- Transposable Elements - model transposons within the sequence
- Repeat regions - will bring a variety of interesting applications


source/Options.py
import os
import pathlib
from types import SimpleNamespace
from source.error_handling import premature_exit, print_and_log


# Constants
# NEAT Path is the path of the main NEAT executable, one level up from where this file resides
NEAT_PATH = pathlib.Path(__file__).resolve().parents[1]


class Options(SimpleNamespace):
    """
    class representing the options
    """

    def __init__(self, config_file=None):
        # For testing purposes, we'll allow a blank options object
        if not config_file:
            self.is_debug = True
        else:
            self.defs = {}
            self.config_file = config_file

            """
            Options flags for gen_reads. This metadata dict gives the type of variable (matching the python types)
            the default value ('.' means no default), and checks. There are four fields: option type (corresponding to
            a python type), default (or None, if no default), criteria 1 and criteria 2. Any items without values 
            should use None as a placeholder.

            Option Type: Current possible valeue are string, int, float, and boolean. These correspond to str, int, 
            float, and bool Python types. For files and directories, use string type.

            Criteria 1: There are two modes of checking: files and numbers. For files, criteria 1 should be set to
            'exists' to check file existence or None to skip a check (as with temp_dir, because that is not a user 
            input. For numbers, criteria 1 should be the lowest acceptable value (inclusive) for that variable.

            Criteria 2: For files, criteria 2 will not be checked, so set to None for consistency. For numbers, this
            should be the highest acceptable value (inclusive).
            (type, default, criteria1 (low/'exists'), criteria2 (high/None))
            """
            arbitrarily_large_number = 1e8
            self.defs['reference'] = ('string', None, 'exists', None)
            self.defs['partition_mode'] = ('string', None, None, None)
            self.defs['read_len'] = ('int', 101, 10, arbitrarily_large_number)
            self.defs['threads'] = ('int', 1, 1, arbitrarily_large_number)
            self.defs['coverage'] = ('float', 10.0, 1, arbitrarily_large_number)
            self.defs['error_model'] = ('string', NEAT_PATH / 'models/errorModel_default.pickle.gz', 'exists', None)
            self.defs['avg_seq_error'] = ('float', None, 0, 0.3)
            self.defs['rescale_qualities'] = ('boolean', False, None, None)
            self.defs['ploidy'] = ('int', 2, 1, 100)
            self.defs['include_vcf'] = ('string', None, 'exists', None)
            self.defs['target_bed'] = ('string', None, 'exists', None)
            self.defs['discard_bed'] = ('string', None, 'exists', None)
            self.defs['off_target_coverage'] = ('float', 0.0, 0, 1)
            self.defs['discard_offtarget'] = ('boolean', False, None, None)
            self.defs['mutation_model'] = ('string', None, 'exists', None)
            self.defs['mutation_rate'] = ('float', None, 0, 0.3)
            self.defs['mutation_bed'] = ('string', None, 'exists', None)
            self.defs['n_handling'] = ('string', None, None, None)

            # Params for cancer (not implemented yet)
            self.defs['cancer'] = ('boolean', False, None, None)
            self.defs['cancer_model'] = ('string', None, 'exists', None)
            self.defs['cancer_purity'] = ('float', 0.8, 0.0, 1.0)

            self.defs['n_cutoff'] = ('int', None, 1, 40)
            self.defs['gc_model'] = ('string', NEAT_PATH / 'models/gcBias_default.pickle.gz', 'exists', None)
            self.defs['paired_ended'] = ('boolean', False, None, None)
            self.defs['fragment_model'] = ('string', None, 'exists', None)
            self.defs['fragment_mean'] = ('float', None, 1, arbitrarily_large_number)
            self.defs['fragment_st_dev'] = ('float', None, 1, arbitrarily_large_number)
            self.defs['produce_bam'] = ('boolean', False, None, None)
            self.defs['produce_vcf'] = ('boolean', False, None, None)
            self.defs['produce_fasta'] = ('boolean', False, None, None)
            self.defs['produce_fastq'] = ('boolean', True, None, None)
            self.defs['force_coverage'] = ('boolean', False, None, None)
            self.defs['temp_dir'] = ('string', pathlib.Path(os.getcwd()), None, None)
            self.defs['debug'] = ('boolean', False, None, None)
            self.defs['rng_value'] = ('int', None, None, None)

            # Cancer options (not yet implemented)
            self.cancer = False
            self.cancer_model = None
            self.cancer_purity = 0.8

            # Read the config file
            self.args = {}
            self.read()
            # Some options checking to clean up the args dict.
            self.check_options()

            self.__dict__.update(self.args)

    def set_value(self, key, value):
        if key in self.defs.keys():
            self.check_and_log_error(key, value, self.defs[key][2], self.defs[key][3])
        self.args[key] = value
        self.__dict__.update(self.args)

    @staticmethod
    def check_and_log_error(keyname, value_to_check, lowval, highval):
        if lowval != "exists" and highval:
            if not (lowval <= value_to_check <= highval):
                print_and_log(f'@{keyname} must be between {lowval} and {highval}.', 'error')
                premature_exit(1)
        elif lowval == "exists":
            if not pathlib.Path(value_to_check).is_file():
                print_and_log(f'The file given to @{keyname} does not exist', 'error')
                premature_exit(1)
        elif not lowval and not highval:
            # This indicates a boolean or dir and we have nothing to check
            pass
        else:
            print_and_log(f'Problem criteria ({lowval, highval}) in Options definitions for {keyname}.', 'critical')
            premature_exit(1)

    def read(self):
        for line in open(self.config_file):
            line = line.strip()
            if line.startswith('@'):
                line_split = [x.strip().strip('@') for x in line.split('=')]
                # If set to a period, then ignore, by convention.
                if line_split[1] == '.':
                    continue
                key = line_split[0]
                # We can ignore any keys users added but haven't coded for. It must be in the defs dict to be examined.
                if key in list(self.defs.keys()):
                    type_of_var, default, criteria1, criteria2 = self.defs[key]
                    # if it's already set to the default value, ignore.
                    if line_split[1] == default:
                        continue

                    # Now we check that the type is correct and it is in range, depending on the type defined for it
                    # If it passes that it gets put into the args dictionary.
                    if type_of_var == 'string':
                        temp = line_split[1]
                        self.check_and_log_error(key, temp, criteria1, criteria2)
                        self.args[key] = temp
                    elif type_of_var == 'int':
                        try:
                            temp = int(line_split[1])
                        except ValueError:
                            print_and_log(f'The value for {key} must be an integer. No decimals allowed.', 'error')
                            premature_exit(1)
                        self.check_and_log_error(key, temp, criteria1, criteria2)
                        self.args[key] = temp
                    elif type_of_var == 'float':
                        try:
                            temp = float(line_split[1])
                        except ValueError:
                            print_and_log(f'The value for {key} must be a float.', 'error')
                            premature_exit(1)
                        self.check_and_log_error(key, temp, criteria1, criteria2)
                        self.args[key] = temp
                    elif type_of_var == 'boolean':
                        try:
                            if line_split[1].lower() == 'true':
                                self.args[key] = True
                            elif line_split[1].lower() == 'false':
                                self.args[key] = False
                            else:
                                raise ValueError
                        except ValueError:
                            print_and_log(f'\nBoolean key @{key} requires a value of "true" or "false" '
                                          f'(case-insensitive).', 'error')
                            premature_exit(1)
                    else:
                        print_and_log(f'BUG: Undefined type in the Options dictionary: {type_of_var}.', 'critical')
                        premature_exit(1)
        # Anything we skipped in the config gets the default value
        # No need to check since these are already CAREFULLY vetted (right!?)
        for key, (_, default, criteria1, criteria2) in self.defs.items():
            if key not in list(self.args.keys()):
                self.args[key] = default

    def check_options(self) -> int:
        """
        Some sanity checks and corrections to the options.
        """

        if self.args['produce_fasta']:
            print_and_log("\nFASTA mode active.", 'info')
            print_and_log("NOTE: At the moment, NEAT can produce a FASTA or FASTQ files, not both.", 'info')
            # Turn off fastq and paired-ended mode for now
            self.args['produce_fastq'] = False
            self.args['paired_ended'] = False
            self.args['fragment_model'] = None
            self.args['fragment_mean'] = None
            self.args['fragment_st_dev'] = None
        if not self.args['produce_bam'] and not self.args['produce_vcf'] \
                and not self.args['produce_fasta'] and not self.args['produce_fastq']:
            print_and_log('No files would be produced, as all file types are set to false', 'error')
            premature_exit(1)
        if not self.args['produce_fastq']:
            print_and_log("Bypassing FASTQ generation.", 'info')
        if self.args['produce_vcf'] and (not self.args['produce_fastq'] and not self.args['produce_bam']
                                         and not self.args['produce_fasta']):
            print_and_log('Only producing VCF output.', 'info')
        if self.args['produce_bam'] and (not self.args['produce_fastq'] and not self.args['produce_vcf']
                                         and not self.args['produce_fasta']):
            print_and_log('Only producing BAM output.', 'info')

        # This next section just checks all the paired ended stuff
        flagged = False
        if self.args['paired_ended']:
            if self.args['fragment_model']:
                self.args['fragment_mean'] = None
                self.args['fragment_st_dev'] = None
            elif self.args['fragment_mean']:
                if not self.args['fragment_st_dev']:
                    flagged = True
            else:
                flagged = True
        if flagged:
            print_and_log("For paired ended mode, you need either a "
                          "@fragment_model or both @fragment_mean and @fragment_st_dev", 'error')
            premature_exit(1)

        self.args['n_handling'] = "ignore"
        if self.args['paired_ended']:
            self.args['n_handling'] = "random"
